src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlTypeWorkarounds.cs

```diff
@@ -61,21 +61,30 @@ internal static DateTime SqlDateTimeToDateTime(int daypart, int timepart)
             const int SQLTicksPerMinute = SQLTicksPerSecond * 60;
             const int SQLTicksPerHour = SQLTicksPerMinute * 60;
             const int SQLTicksPerDay = SQLTicksPerHour * 24;
-            const int MinDay = -53690;                // Jan 1 1753
-            const int MaxDay = 2958463;               // Dec 31 9999 is this many days from Jan 1 1900
-            const int MinTime = 0;                    // 00:00:0:000PM
-            const int MaxTime = SQLTicksPerDay - 1; // = 25919999,  11:59:59:997PM
-
-            if (daypart < MinDay || daypart > MaxDay || timepart < MinTime || timepart > MaxTime)
+            //const int MinDay = -53690;                // Jan 1 1753
+            const uint MinDayOffset = 53690;            // postive value of MinDay used to pull negative values up to 0 so a single check can be used
+            const uint MaxDay = 2958463;               // Dec 31 9999 is this many days from Jan 1 1900
+            const uint MaxTime = SQLTicksPerDay - 1; // = 25919999,  11:59:59:997PM
+            const long BaseDateTicks = 599266080000000000L;//new DateTime(1900, 1, 1).Ticks;
+
+            // casting to uint wraps negative values to large positive ones above the valid 
+            // ranges so the lower bound doesn't need to be checked
+            if ((uint)(daypart + MinDayOffset) > (MaxDay + MinDayOffset) || (uint)timepart > MaxTime)
             {
-                throw new OverflowException(SQLResource.DateTimeOverflowMessage);
+                ThrowOverflowException();
             }
 
-            long baseDateTicks = new DateTime(1900, 1, 1).Ticks;
             long dayticks = daypart * TimeSpan.TicksPerDay;
-            long timeticks = ((long)(timepart / SQLTicksPerMillisecond + 0.5)) * TimeSpan.TicksPerMillisecond;
+            double timePartPerMs = timepart / SQLTicksPerMillisecond;
+            timePartPerMs += 0.5;
+            long timeTicks = ((long)timePartPerMs) * TimeSpan.TicksPerMillisecond;
+            long totalTicks = BaseDateTicks + dayticks + timeTicks;
+            return new DateTime(totalTicks);
+        }
 
-            return new DateTime(baseDateTicks + dayticks + timeticks);
+        private static void ThrowOverflowException()
+        {
+            throw new OverflowException(SQLResource.DateTimeOverflowMessage);
         }
```

Comment:
Could you use utility class to gather the exception message, please?

Reply:
Any particular reason? and any particular place you want it to be?

Reply:
To prevent redundancy and have better management over them. You can use the `SqlUtil.cs` file in the common source folder.
FYI, we already have three `SqlUtil.cs` files, one for each project plus what I mentioned.

Reply:
Done. I'm not convinced that it's a great idea because it seems to be blurring the distinction between a couple of namespaces. It's also important for the codegen that the exception be thrown from a helper method so it just adds another level of indirection that doesn't seem useful.

Reply:
It seems you included it in the wrong file. As I mentioned, there are three **SqlUtil.cs** files that one of them is in the common folder to use for both projects that it's a better choice for your case. _You can find it if you pull the latest source code._

Also, we usually create a specific new exception object and use it where it's required to throw (don't throw an exception in **SqlUtil**). I don't think you need the private local static method (**ThrowOverflowException**), you can throw `SQL.DateTimeOverflow()` where you need it.

I found `SqlUtil.UDTInvalidSize()` method has the same problem; could you modify it in the same way, please?

About the points, you mentioned we can discuss more in a separate session to find a better way. Until then, keeping the source unify prevents ambiguity.

Reply:
> I don't think you need the private local static method (ThrowOverflowException), you can throw SQL.DateTimeOverflow() where you need it.

No, I do need it because throwing inline has a significant performance penalty. Use sharplab or disasmo and check the tier1 assembly that's generated. A call to a throw helper function is faster than doing it inline. Or spin up a bdn benchmark like I did and see what changes make a difference. This one is the best I can get it, even tiny changes have big differences.

There really shouldn't be multiple SqlUtil files and whatever they contain should be split out into correctly named files. I can move around the things for this PR but other changes will have to be done in the larger merge process or in individual PR's.

Reply:
Since we still don't have a perf lab, clarify your reasons by adding some notes to avoid a performance regression hereafter, please.

Reply:
NoInlining hint added (though it's currently not necessary) and a comment about why the method is extracted from it's caller added for future users.

Reply:
I've also added the DateTimeOverflow method in netfx. In this source there are only two SqlUtil files there aren't conflicts. I have seen that in the main branch there is a shared one but it also comes with a compiler warning that indicates one of them is being ignored, so I don't think that's doing whatever it was intended to do.

Reply:
Consider the difference between my machine configuration and Warith in result values.
Don't think I'm jealous of you @Wraith2 at all. 😉 

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-8665U CPU 1.90GHz (Coffee Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.201
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  DefaultJob : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT


```
|                  Method |     Mean |    Error |   StdDev | Ratio |
|------------------------ |---------:|---------:|---------:|------:|
|                Original | 481.6 ms | 12.74 ms | 36.76 ms |  1.00 |
|                     New | 130.2 ms |  1.76 ms |  1.81 ms |  0.27 |
| NewUtilPrivateException | 136.0 ms |  2.95 ms |  8.22 ms |  0.28 |
|  NewUtilInlineException | 145.1 ms |  3.28 ms |  9.47 ms |  0.30 |

- The benchmark code based on Wraith's code:  https://gist.github.com/DavoudEshtehari/267fb6cc6dd09475f51a41ed48364f17